service/matching/pri_task_reader.go

```diff
@@ -159,11 +159,11 @@ func (tr *priTaskReader) completeTask(task *internalTask, res taskResponse) {
 	}
 
 	tr.lock.Lock()
+	defer tr.lock.Unlock()
```

Comment:
There's a race causing the `should not move backwards ` assertion failure if the unlock happens before the call to `updateAckLevelAndCount`.

Reply:
Oh, duh, actually it doesn't matter if we take db lock inside taskReader lock or not, since we do take it in this function. So maybe we can leave this here.

I do want to fix this later. Also the new pri- implementations do have one regression from the old ones in that the stats+status api calls block on db.lock, where before they just read atomics. I don't think that's a big deal but it should also be fixed eventually.

Reply:
I was completely with you on your first comment; but the second one I can't quite follow. Why doesn't it actually matter if we still hold the taskReader lock while also holding the db lock?

Reply:
In general, if they can be independent they should be. But I agree in this case it doesn't really matter.

I think the db lock should be reworked, maybe including this one too. There's two things I would like to be true:

1. tr.completeTask (which is in the poll path) never has to wait for db ops.
2. The stats/status calls  (tr.getLevels, tr.getLoadedTasks, db.getApproximateBacklogCount) never have to wait for db ops.

I think 2 is mostly true on main but regressed on the branch. I think that's okay though, they're just informative calls. 1 isn't true now but it should be, and with all the refactoring, we're closer to making it true.